### PR TITLE
Format simulation (ready for review)

### PR DESCRIPTION
**Pull request by thecharlieblake**
Note: now fully ready from my side (rebased onto latest main and docstrings added)

### CODE REVIEW

File: unit_scaling/transforms/_simulate_format.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        fwd_format (FPFormat): the quantisation format to be used in the forward pass
+            (activations and weights)
+        bwd_format (FPFormat): the quantisation format to be used in the backward pass
+            (gradients of activations and weights)
+
+    Returns:
+        nn.Module: a new module which when used, will run using the simulated formats.
+    """
+    module = deepcopy(module)
+    torch._dynamo.reset()  # type: ignore[no-untyped-call]
```
Comment (DouglasOrr) [inline unit_scaling/transforms/_simulate_format.py:162]: What happens if you don't do that?

File: unit_scaling/transforms/utils.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            ).add_options(self)
+    return super(
+        torch._dynamo.variables.functions.UserMethodVariable, self
+    ).call_function(tx, args, kwargs)
+
+
+@contextmanager
+def _expand_modules_patch(non_recurse_functions):  # type: ignore[no-untyped-def]
+    patcher_a = patch(
+        "torch._dynamo.allowed_functions._allowed_function_ids",
```
Comment (DouglasOrr) [inline unit_scaling/transforms/utils.py:54]: This is very likely to break in future PyTorch versions, but should tell us when it does!

**Review by DouglasOrr — APPROVED** [review]
Optional comments/questions, but in summary, LGTM :shipit:. Impressive graph manipulation bits!

File: unit_scaling/formats.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        """The minimum absolute normal value representable by the format."""
+        min_exponent = 1 - 2 ** (self.exponent_bits - 1)
+        return cast(float, 2**min_exponent)
+
+    @property
+    def min_absolute_subnormal(self) -> float:
+        """The minimum absolute subnormal value representable by the format."""
+        return self.min_absolute_normal * 2.0**-self.mantissa_bits
+
+    def quantise_no_grad(self, x: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (thecharlieblake) [reply unit_scaling/formats.py:69]: Agreed & fixed!

File: unit_scaling/transforms/_simulate_format.py (hunk 1 above)
Comment (thecharlieblake) [reply unit_scaling/transforms/_simulate_format.py:162]: Torch dynamo throws a warning if you call it with a different backend without calling that first, and may (I'm not sure) not actually use the new backend. Not sure what's going on there, but I don't think this hurts in our case.

File: unit_scaling/transforms/utils.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        new=_get_patched_allowed_function_ids(non_recurse_functions),
+    )
+    patcher_b = patch(
+        "torch._dynamo.variables.functions.UserMethodVariable.call_function",
+        new=_patched_call_function,
+    )
+    try:
+        yield (patcher_a.start(), patcher_b.start())
+    finally:
+        patcher_a.stop()
```
[the remark this answers is not in the document]
Comment (thecharlieblake) [reply unit_scaling/transforms/utils.py:64]: Neat! Updated accordingly

**Pull request merged**